File: decision-engine/service/intelligence.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package service

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"net/http"
	"regexp"
	"sort"

	"github.com/e6qu/intraktible/decision-engine/domain"
	"github.com/e6qu/intraktible/decision-engine/events"
	"github.com/e6qu/intraktible/decision-engine/flows"
	"github.com/e6qu/intraktible/decision-engine/history"
	"github.com/e6qu/intraktible/decision-engine/policy"
	"github.com/e6qu/intraktible/platform/httpx"
	"github.com/e6qu/intraktible/platform/identity"
)

// nodeStatsScan caps how many recent completed decisions a node-stats request
// aggregates — the read model has no field index, so the scan is bounded.
const nodeStatsScan = 500

// deriveDisposition assigns a disposition over a run's output by applying the
// flow's bound policy, reusing the exact lookup+Apply path the decide handler
// uses (policy.ActiveForFlow → ver.Spec.Apply). No policy bound → empty
// disposition; a policy that cannot evaluate refers (matching decide), so a
// re-run never fails on a policy problem. Only a store error is returned.
func (s *Service) deriveDisposition(ctx context.Context, id identity.Identity, slug string, output map[string]any) (policy.Disposition, error) {
	_, ver, ok, err := policy.ActiveForFlow(ctx, s.store, id, slug)
	if err != nil {
		return "", err
	}
	if !ok {
		return "", nil
	}
	out, applyErr := ver.Spec.Apply(output)
	// A policy that can't evaluate this output falls back to a safe manual referral
	// rather than aborting the probe (structured so the nil return isn't masking err).
	disp := policy.Refer
	if applyErr == nil {
		disp = out.Disposition
	}
	return disp, nil
}

// publishedGraph returns the flow's published graph for the requested version (0
// = latest published). It is the same projection-sourced graph decide/backtest
// run, so a re-run matches the live decision path.
func publishedGraph(fv flows.FlowView, version int) (events.Graph, bool) {
	if len(fv.Versions) == 0 {
		return events.Graph{}, false
	}
	g, err := flows.GraphForVersion(fv, version)
	if err != nil {
		return events.Graph{}, false
	}
	return g, true
}

type nodeStat struct {
	NodeID string          `json:"node_id"`
	Type   events.NodeType `json:"type"`
	Count  int             `json:"count"`
	Pct    float64         `json:"pct"`
}

type nodeStatsResponse struct {
	Total        int            `json:"total"`
	Dispositions map[string]int `json:"dispositions"`
	Nodes        []nodeStat     `json:"nodes"`
}

// nodeStats is the traversal heatmap: across the flow's recent COMPLETED
// decisions, how many hit each node of its current published graph (a node is
// "hit" when it appears in a record's trace), plus the disposition tally.
//
//	GET /v1/flows/{flow_id}/node-stats?environment=production
func (s *Service) nodeStats(w http.ResponseWriter, r *http.Request) {
	id, ok := httpx.Caller(w, r)
	if !ok {
		return
	}
	fv, found, err := flows.Read(r.Context(), s.store, id, r.PathValue("flow_id"))
	if err != nil {
		httpx.Error(w, http.StatusInternalServerError, err)
		return
	}
	if !found {
		httpx.Error(w, http.StatusNotFound, fmt.Errorf("flow not found"))
		return
	}
	graph, ok := publishedGraph(fv, 0)
	if !ok {
		// An unpublished flow has no graph to attribute hits to: zero stats, not an error.
		httpx.JSON(w, http.StatusOK, nodeStatsResponse{
			Dispositions: map[string]int{"approve": 0, "decline": 0, "refer": 0},
			Nodes:        []nodeStat{},
		})
		return
	}

	page, err := history.ListPage(r.Context(), s.store, id, history.Filter{
		Slug: fv.Slug, Environment: r.URL.Query().Get("environment"), Status: "completed", Limit: nodeStatsScan,
	})
	if err != nil {
		httpx.Error(w, http.StatusInternalServerError, err)
		return
	}

	httpx.JSON(w, http.StatusOK, aggregateNodeStats(graph, page.Records))
}

// aggregateNodeStats tallies node hits and the disposition distribution across a set
// of completed-decision records, attributing hits to the nodes of the current
// published graph. A record's node list may reference ids no longer in the graph
// (graph evolved since the decision ran) — those are counted internally but only the
// current graph's nodes are reported, so the per-node pct stays in [0,1]. Pure and
// bounded by the record/graph sizes the caller already capped.
func aggregateNodeStats(graph events.Graph, records []history.Record) nodeStatsResponse {
	counts := make(map[string]int, len(graph.Nodes))
	dispositions := map[string]int{"approve": 0, "decline": 0, "refer": 0}
	for _, rec := range records {
		seen := make(map[string]bool, len(rec.Nodes))
		for _, n := range rec.Nodes {
			if seen[n.NodeID] {
				continue
			}
			seen[n.NodeID] = true
			counts[n.NodeID]++
		}
		if _, tracked := dispositions[rec.Disposition]; tracked {
			dispositions[rec.Disposition]++
		}
	}

	total := len(records)
	nodes := make([]nodeStat, 0, len(graph.Nodes))
	for _, n := range graph.Nodes {
		c := counts[n.ID]
		// A node can be hit at most once per record (deduped above), so c <= total and
		// pct stays in [0,1]; total>0 guards the division.
		var pct float64
		if total > 0 {
			pct = float64(c) / float64(total)
		}
		nodes = append(nodes, nodeStat{NodeID: n.ID, Type: n.Type, Count: c, Pct: pct})
	}
	return nodeStatsResponse{Total: total, Dispositions: dispositions, Nodes: nodes}
}

// dispositionRank orders dispositions by favorability (higher is more favorable),
// so the counterfactual search only reports a change that strictly improves the
// outcome. An unknown/empty disposition is least favorable.
func dispositionRank(d policy.Disposition) int {
	switch d {
	case policy.Approve:
		return 2
	case policy.Refer:
		return 1
	case policy.Decline:
		return 0
	default:
		return -1
	}
}

// counterfactual caps. cfEvalsPerField bounds the binary search per numeric field;
// cfMaxEvals bounds the whole request so a wide input can't run unbounded.
const (
	cfEvalsPerField = 24
	cfMaxEvals      = 400
)

type flip struct {
	Field       string  `json:"field"`
	From        float64 `json:"from"`
	To          float64 `json:"to"`
	Direction   string  `json:"direction"`
	Disposition string  `json:"disposition"`
}

type counterfactualResponse struct {
	Disposition string `json:"disposition"`
	Flips       []flip `json:"flips"`
	Searched    int    `json:"searched"`
}

// counterfactual answers "what would flip this?": for an unfavorable decision, it
// searches each numeric input field for the minimal single-field change that
// re-runs to a more favorable disposition. It re-runs with domain.Execute then
// applies the same policy derivation the decide path uses, so a reported flip is
// a real disposition change. Pure and bounded (no recording, capped evals).
//
//	POST /v1/decisions/{id}/counterfactual
func (s *Service) counterfactual(w http.ResponseWriter, r *http.Request) {
	id, ok := httpx.Caller(w, r)
	if !ok {
		return
	}
	rec, found, err := history.Read(r.Context(), s.store, id, r.PathValue("id"))
	if err != nil {
		httpx.Error(w, http.StatusInternalServerError, err)
		return
	}
	if !found {
		httpx.Error(w, http.StatusNotFound, fmt.Errorf("decision not found"))
		return
	}
	original := policy.Disposition(rec.Disposition)
	if original == policy.Approve {
		// Already favorable: nothing to flip.
		httpx.JSON(w, http.StatusOK, counterfactualResponse{Disposition: "approve", Flips: []flip{}, Searched: 0})
		return
	}

	fv, found, err := flows.Read(r.Context(), s.store, id, rec.FlowID)
	if err != nil {
		httpx.Error(w, http.StatusInternalServerError, err)
		return
	}
	if !found {
		httpx.Error(w, http.StatusNotFound, fmt.Errorf("flow not found"))
		return
	}
	graph, err := flows.GraphForVersion(fv, rec.Version)
	if err != nil {
		httpx.Error(w, http.StatusBadRequest, err)
		return
	}

	var data map[string]any
	if len(rec.Data) > 0 {
		if err := json.Unmarshal(rec.Data, &data); err != nil {
			httpx.Error(w, http.StatusBadRequest, fmt.Errorf("decision input is not a JSON object: %w", err))
			return
		}
	}
	if data == nil {
		data = map[string]any{}
	}
	// The recorded input already carries any pre-resolved engine namespaces
	// (features/connect/ai/predict). Re-running over the same map keeps those nodes
	// from failing, since the pure re-run has no provider to re-resolve them.

	searched := 0
	var flips []flip
	// A field the graph recomputes (an assignment/scorecard/model target) is not a lever
	// the applicant controls, so it must not be offered as a counterfactual.
	derived := graphTargets(graph)
	for _, field := range numericFields(data) {
		if derived[field] {
			continue
		}
		if searched >= cfMaxEvals {
			break
		}
		budget := cfMaxEvals - searched
		if budget > cfEvalsPerField {
			budget = cfEvalsPerField
		}
		f, used := s.searchFlip(r.Context(), id, fv.Slug, graph, data, field, original, budget)
		searched += used
		if f != nil {
			flips = append(flips, *f)
		}
	}

	// Smallest relative change first, so the closest-to-the-threshold lever leads.
	sort.SliceStable(flips, func(i, j int) bool {
		return relChange(flips[i]) < relChange(flips[j])
	})
	if flips == nil {
		flips = []flip{}
	}
	httpx.JSON(w, http.StatusOK, counterfactualResponse{
		Disposition: string(original), Flips: flips, Searched: searched,
	})
}

// relChange is a flip's magnitude relative to the original value, used to order
// flips by which lever moves least.
func relChange(f flip) float64 {
	denom := math.Abs(f.From)
	if denom == 0 {
		denom = 1
	}
	return math.Abs(f.To-f.From) / denom
}

// roundCF rounds a counterfactual target for display + degenerate detection: whole
// numbers for large magnitudes, two decimals otherwise.
func roundCF(v float64) float64 {
	if math.Abs(v) >= 100 {
		return math.Round(v)
	}
	return math.Round(v*100) / 100
}

var cfTargetRe = regexp.MustCompile(`"(?:target|output)"\s*:\s*"([^"]*)"`)

// graphTargets is the set of fields the graph assigns/derives (assignment/scorecard/model
// outputs); they must not be counterfactual levers since the graph recomputes them
// regardless of the supplied input.
func graphTargets(graph events.Graph) map[string]bool {
	out := map[string]bool{}
	for _, n := range graph.Nodes {
		for _, m := range cfTargetRe.FindAllStringSubmatch(string(n.Config), -1) {
			out[m[1]] = true
		}
	}
	return out
}

// searchFlip looks for the smallest single-field change to `field` that re-runs to
// a strictly more favorable disposition than `original`. It binary-searches both
// directions over [value/8 .. value*8] (with a small absolute floor so a near-zero
// value still has room to move) and returns the smaller |to-from| flip found, plus
// the number of evaluations used (bounded by budget).
func (s *Service) searchFlip(ctx context.Context, id identity.Identity, slug string, graph events.Graph, data map[string]any, field string, original policy.Disposition, budget int) (*flip, int) {
	val := data[field].(float64)
	// A non-finite field value has no meaningful threshold to search toward and would
	// poison the span/direction math (NaN comparisons, Inf±Inf), so it is not a lever.
	if math.IsNaN(val) || math.IsInf(val, 0) {
		return nil, 0
	}
	used := 0
	eval := func(x float64) (policy.Disposition, bool) {
		if used >= budget {
			return "", false
		}
		used++
		perturbed := make(map[string]any, len(data))
		for k, v := range data {
			perturbed[k] = v
		}
		perturbed[field] = x
		run := domain.Execute(graph, perturbed)
		if run.Status != domain.StatusCompleted {
			return "", true
		}
		disp, err := s.deriveDisposition(ctx, id, slug, run.Output)
		if err != nil {
			return "", false
		}
		return disp, true
	}

	originalRank := dispositionRank(original)
	better := func(d policy.Disposition) bool { return dispositionRank(d) > originalRank }

	// Bound the search to a sensible window — [val/4, val*4] for a positive field — so it
	// never suggests a negative income, a zero denominator, or a wild change. hi stays
	// finite for an extreme magnitude (val near MaxFloat64 would otherwise overflow).
	var lo, hi float64
	if val > 0 {
		lo = val * 0.125
		hi = val * 8
	} else {
		pad := math.Abs(val) * 8
		if pad < 25 {
			pad = 25
		}
		lo = math.Max(0, val-pad)
		hi = val + pad
	}
	if hi > math.MaxFloat64/2 {
		hi = math.MaxFloat64 / 2
	}

	up := boundFlip(eval, val, hi, better)
	down := boundFlip(eval, val, lo, better)

	best := closestFlip(val, up, down)
	if best == nil {
		return nil, used
	}
	to := roundCF(best.to)
	if to == roundCF(val) { // boundary sits on the current value — not a useful lever
		return nil, used
	}
	dir := "increase"
	if to < val {
		dir = "decrease"
	}
	return &flip{
		Field: field, From: val, To: to, Direction: dir, Disposition: string(best.disp),
	}, used
}

// flipPoint is a found favorable perturbation: the field value and the disposition it yields.
type flipPoint struct {
	to   float64
	disp policy.Disposition
}

// boundFlip binary-searches between `from` (assumed not better) and `to` for the
// value closest to `from` that evaluates to a strictly better disposition. It
// returns nil when `to` itself is not better (no flip in this direction) or the
// eval budget runs out.
func boundFlip(eval func(float64) (policy.Disposition, bool), from, to float64, better func(policy.Disposition) bool) *flipPoint {
	endDisp, ok := eval(to)
	if !ok || !better(endDisp) {
		return nil
	}
	lo, hi := from, to
	bestTo, bestDisp := to, endDisp
	// A fixed iteration count converges the threshold without unbounded evals; the
	// eval budget cuts it shorter if exhausted.
	for i := 0; i < 16; i++ {
		mid := (lo + hi) / 2
		disp, ok := eval(mid)
		if !ok {
			break
		}
		if better(disp) {
			bestTo, bestDisp = mid, disp
			hi = mid
		} else {
			lo = mid
		}
	}
	return &flipPoint{to: bestTo, disp: bestDisp}
}

// closestFlip returns whichever of two candidate flips moves the field least from val.
func closestFlip(val float64, a, b *flipPoint) *flipPoint {
	switch {
	case a == nil:
		return b
	case b == nil:
		return a
	case math.Abs(a.to-val) <= math.Abs(b.to-val):
		return a
	default:
		return b
	}
}

// numericFields returns the sorted top-level numeric input field names (engine
// namespaces are skipped — they are pre-resolved context, not levers a caller controls).
func numericFields(data map[string]any) []string {
	var fields []string
	for k, v := range data {
		if isReservedNamespace(k) {
			continue
		}
		if _, ok := v.(float64); ok {
			fields = append(fields, k)
		}
	}
	sort.Strings(fields)
	return fields
}

func isReservedNamespace(k string) bool {
	switch k {
	case "features", "connect", "ai", "predict":
		return true
	default:
		return false
	}
}

File: web/src/lib/demo/intelligence.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Demo-side compute for the "decision intelligence" features — node-traversal stats
// (heatmap), counterfactual search ("what would flip this?"), and coverage/red-team
// fuzzing. The static demo has no Go backend, so these mirror the real endpoints by
// re-running the demo engine: runFlow is pure (it threads a record through the graph
// and never touches the store), so re-running it hundreds of times is safe and the
// results are genuine rather than canned.

import type {
  Flow,
  FlowGraph,
  Decision,
  Disposition,
  FlowNodeStats,
  Counterfactual,
  CounterfactualFlip,
  Coverage
} from '$lib/api';
import { runFlow, pickVersion } from './engine';

type DispCount = { approve: number; decline: number; refer: number };
function emptyDisp(): DispCount {
  return { approve: 0, decline: 0, refer: 0 };
}
function tally(d: DispCount, disp?: Disposition): void {
  if (disp === 'approve' || disp === 'decline' || disp === 'refer') d[disp] += 1;
}

// nodeStats aggregates how often each node of the flow's latest graph was traversed
// across its recorded decisions — the data behind the heatmap (hot vs never-hit nodes).
export function nodeStats(flow: Flow, decisions: Decision[]): FlowNodeStats {
  const decs = decisions.filter((d) => d.flow_id === flow.flow_id && d.status === 'completed');
  const { graph } = pickVersion(flow, 'production');
  const counts = new Map<string, number>();
  const dispositions = emptyDisp();
  for (const d of decs) {
    for (const n of d.nodes ?? []) counts.set(n.node_id, (counts.get(n.node_id) ?? 0) + 1);
    tally(dispositions, d.disposition);
  }
  const total = decs.length;
  const nodes = graph.nodes.map((n) => {
    const count = counts.get(n.id) ?? 0;
    return { node_id: n.id, type: n.type, count, pct: total ? count / total : 0 };
  });
  return { total, dispositions, nodes };
}

// --- Counterfactual: minimal single-field change that flips a non-favorable decision ---

// Dispositions ordered worst -> best, so we can tell whether a re-run improved.
const RANK: Record<string, number> = { decline: 0, refer: 1, approve: 2 };
function isBetter(next: Disposition | undefined, base: Disposition): boolean {
  return next != null && (RANK[next] ?? -1) > (RANK[base] ?? -1);
}

export function counterfactual(flow: Flow, decision: Decision): Counterfactual {
  const base = decision.disposition;
  if (!base || base === 'approve') {
    return { disposition: base ?? 'approve', flips: [], searched: 0 };
  }
  const { graph } = pickVersion(flow, decision.environment ?? 'production');
  const data = (decision.data as Record<string, unknown>) ?? {};
  const numeric = Object.entries(data)
    .filter(([, v]) => typeof v === 'number' && Number.isFinite(v as number))
    .map(([k]) => k);
  let searched = 0;
  const disp = (val: number, field: string): Disposition | undefined => {
    searched += 1;
    return runFlow(flow, graph, { ...data, [field]: val }).disposition;
  };
  const flips: CounterfactualFlip[] = [];
  for (const field of numeric) {
    if (searched > 360) break;
    const flip = searchFlip(field, data[field] as number, base, disp);
    if (flip) flips.push(flip);
  }
  // Smallest relative change first — the easiest lever to pull.
  flips.sort((a, b) => relChange(a) - relChange(b));
  return { disposition: base, flips, searched };
}

function relChange(f: CounterfactualFlip): number {
  return Math.abs((f.to - f.from) / (f.from || 1));
}

// searchFlip scans a field outward in both directions for the nearest value that
// improves the disposition, then binary-searches the boundary. ~22 evals per field.
function searchFlip(
  field: string,
  from: number,
  base: Disposition,
  disp: (val: number, field: string) => Disposition | undefined
): CounterfactualFlip | null {
  const span = Math.max(Math.abs(from) * 8, 50);
  const STEPS = 8;
  for (let i = 1; i <= STEPS; i++) {
    for (const dir of [1, -1]) {
      const val = from + (dir * span * i) / STEPS;
      const d = disp(val, field);
      if (isBetter(d, base)) {
        const r = refineBoundary(from, val, base, field, disp);
        return {
          field,
          from,
          to: round(r.to),
          direction: r.to > from ? 'increase' : 'decrease',
          disposition: r.d
        };
      }
    }
  }
  return null;
}

// refineBoundary narrows [from (not improving) .. found (improving)] to the closest
// value that still improves the disposition.
function refineBoundary(
  from: number,
  found: number,
  base: Disposition,
  field: string,
  disp: (val: number, field: string) => Disposition | undefined
): { to: number; d: Disposition } {
  let a = from;
  let b = found;
  let dB = disp(b, field) as Disposition;
  for (let i = 0; i < 12; i++) {
    const mid = (a + b) / 2;
    const d = disp(mid, field);
    if (isBetter(d, base)) {
      b = mid;
      dB = d as Disposition;
    } else {
      a = mid;
    }
  }
  return { to: b, d: dB };
}

function round(v: number): number {
  return Math.abs(v) >= 100 ? Math.round(v) : Math.round(v * 100) / 100;
}

// --- Coverage / red-team: fuzz synthetic inputs and report node/branch coverage --------

export function coverage(flow: Flow, graph: FlowGraph, runs: number): Coverage {
  const n = Math.max(1, Math.min(runs || 200, 2000));
  const { fields, thresholds } = discoverInputs(graph);
  const nodeHits = new Map<string, number>();
  const edgeHits = new Map<string, number>();
  const dispositions = emptyDisp();
  // The branch is part of the key: a split has several edges with the same from/to but
  // different conditions, and only the edge whose condition the run took is hit.
  const edgeKey = (from: string, to: string, branch: string) => `${from}>${to}>${branch}`;
  for (let i = 0; i < n; i++) {
    const r = runFlow(flow, graph, synthInput(fields, thresholds, i));
    const trace = r.nodes ?? [];
    for (const nr of trace) nodeHits.set(nr.node_id, (nodeHits.get(nr.node_id) ?? 0) + 1);
    // Consecutive nodes name the edge traversed; a split records its chosen branch in its
    // output, which disambiguates parallel edges between the same two nodes.
    for (let j = 0; j + 1 < trace.length; j++) {
      const chosen = (trace[j].output as { branch?: string } | undefined)?.branch ?? '';
      const k = edgeKey(trace[j].node_id, trace[j + 1].node_id, chosen);
      edgeHits.set(k, (edgeHits.get(k) ?? 0) + 1);
    }
    tally(dispositions, r.disposition);
  }
  const nodes = graph.nodes.map((nd) => ({
    node_id: nd.id,
    type: nd.type,
    hits: nodeHits.get(nd.id) ?? 0
  }));
  const hitsFor = (e: { from: string; to: string; branch?: string }) =>
    edgeHits.get(edgeKey(e.from, e.to, e.branch ?? '')) ?? 0;
  const branches = graph.edges.map((e) => ({
    from: e.from,
    to: e.to,
    branch: e.branch ?? '',
    hits: hitsFor(e)
  }));
  const dead_nodes = nodes.filter((nd) => nd.hits === 0).map((nd) => nd.node_id);
  const dead_branches = graph.edges
    .filter((e) => e.branch && hitsFor(e) === 0)
    .map((e) => ({ from: e.from, to: e.to, branch: e.branch ?? '' }));
  return { runs: n, fields, nodes, branches, dispositions, dead_nodes, dead_branches };
}

// RESERVED also excludes the node namespaces (predict.x / connect.x / ai.x are derived
// outputs, not inputs the caller supplies), so the bare 'predict'/'connect'/'ai' that a
// dotted reference splits into don't leak in as fuzz dimensions.
const RESERVED = new Set([
  'true',
  'false',
  'null',
  'and',
  'or',
  'not',
  'if',
  'else',
  'min',
  'max',
  'abs',
  'round',
  'floor',
  'ceil',
  'predict',
  'connect',
  'ai'
]);

// discoverInputs harvests both the input identifiers a graph branches on (names that
// appear in expressions / conditions but are NOT themselves assigned — a target/output is
// derived) and the numeric thresholds those conditions compare against, so the fuzzer can
// sample right around each gate. Heuristic but matches what the real coverage endpoint does.
function discoverInputs(graph: FlowGraph): { fields: string[]; thresholds: number[] } {
  const found = new Set<string>();
  const targets = new Set<string>();
  const derived = new Set<string>(); // suffixes of a dotted ref (predict.probability) — not inputs
  const thresholds = new Set<number>();
  const harvest = (s: string) => {
    for (const m of s.matchAll(/[a-zA-Z_][a-zA-Z0-9_.]*\.([a-zA-Z_][a-zA-Z0-9_]*)/g)) {
      derived.add(m[1]);
    }
    for (const m of s.matchAll(/[a-zA-Z_][a-zA-Z0-9_]*/g)) {
      if (!RESERVED.has(m[0])) found.add(m[0]);
    }
    for (const m of s.matchAll(/-?\d+(?:\.\d+)?/g)) {
      const v = Number(m[0]);
      if (Number.isFinite(v)) thresholds.add(v);
    }
  };
  for (const e of graph.edges) if (e.branch) harvest(e.branch);
  for (const nd of graph.nodes) {
    const cfg = JSON.stringify(nd.config ?? {});
    for (const m of cfg.matchAll(/"(?:expr|when)"\s*:\s*"([^"]*)"/g)) harvest(m[1]);
    for (const m of cfg.matchAll(/"(?:target|output)"\s*:\s*"([^"]*)"/g)) targets.add(m[1]);
  }
  const fields = [...found]
    .filter((f) => !targets.has(f) && !derived.has(f) && !f.includes('.'))
    .slice(0, 12);
  return { fields, thresholds: [...thresholds] };
}

// rangeFor scales a synthetic value to the field's likely magnitude so score gates and
// money gates both get exercised.
function rangeFor(field: string): number {
  if (/income|amount|balance|limit|revenue|salary|loan|debt/i.test(field)) return 100_000;
  if (/fico|credit_score/i.test(field)) return 850;
  return 100;
}

// synthInput deterministically generates one fuzz input from an index (no Math.random,
// so the endpoint is reproducible — the same flow + runs gives the same coverage). Half
// the time a field is sampled near one of the graph's own thresholds (so both sides of
// each gate get hit — better branch coverage and a realistic disposition spread); the
// rest of the time it's a broad random value over the field's plausible range.
function synthInput(fields: string[], thresholds: number[], seed: number): Record<string, number> {
  let s = ((seed + 1) * 2654435761) % 2147483647;
  const rnd = () => {
    s = (s * 1103515245 + 12345) & 0x7fffffff;
    return s / 0x7fffffff;
  };
  const out: Record<string, number> = {};
  for (const f of fields) {
    if (thresholds.length && rnd() < 0.5) {
      const t = thresholds[Math.floor(rnd() * thresholds.length)];
      const jitter = (rnd() - 0.5) * Math.max(Math.abs(t) * 0.6, 4);
      out[f] = Math.round((t + jitter) * 100) / 100;
    } else {
      out[f] = Math.round(rnd() * rangeFor(f) * 100) / 100;
    }
  }
  return out;
}
